esp-hal/src/i2c/master/mod.rs

```diff
@@ -521,33 +619,38 @@ where
     }
 
     /// Writes bytes to slave with address `address`
-    pub fn write(&mut self, address: u8, buffer: &[u8]) -> Result<(), Error> {
+    pub fn write<A: AddressMode>(&mut self, address: A, buffer: &[u8]) -> Result<(), Error> {
```

Comment:
So with this way of doing things, when we add 10bit support, won't this break code like this:

```rust
i2c.write(0xaa /* <- ambiguous type here, could be a u8 or a u16 */, &[])?;
```

Reply:
We would be able to implement for `i32`[^1] and decide what to do with that. But it would probably be unclear what i32 would represent. Instead, I would recommend implementing AddressMode for `u8` and `TenBit(u16)` instead, not for other numeric types.

[^1]: https://rust.godbolt.org/z/Wfjh7PMMc

Reply:
oh yeah - we can't have defaults for type parameters on functions :( which would have been an easy solution

> I would recommend implementing AddressMode for u8 and TenBit(u16) instead, not for other numeric types

I like this solution a lot - 10-bit is uncommon enough to have a new-type for it

Reply:
Based on these ideas, what about the following?

* drop the trait (meaning we don't need the private module too)
* Make a public enum `I2cAddress` which has 7 bit and 10 bit variants
* make our public APIs take `Into<I2cAddress>`
* Add a singular `From<u8> for I2cAddress`

Doing it this way means a user can always pass `I2cAddress` directly, but as a convenience, we also allow a single `u8` address.